refactor: switch to NPM workspaces

not-my-profile opened the pull request:
The `lerna bootstrap` command is deprecated[1]
and resulted in a very weird CI failure (#685).

This commit switches the repository to instead use
NPM workspaces.[2]

[1]: https://github.com/lerna/lerna/discussions/3410
[2]: https://docs.npmjs.com/cli/v7/using-npm/workspaces

Fixes #685.

stephenmathieson commented on packages/reporter-earl/tests/axeReporterEarl.test.ts:
```diff
@@ -4,6 +4,7 @@ import { getDummyData } from './utils';
 import axeReporterEarl, { createEarlReport } from '../src/axeReporterEarl';
 import * as context from '../src/context.json';
 import { RawResult, EarlType } from '../src/types';
+import { describe, beforeEach, test, expect } from '@jest/globals';
```
It looks like the build is failing due to Jest related dependency issues.

stephenmathieson replied:
That's odd. I was not aware of that. Thanks 👍

not-my-profile commented:
Ok I got it working ... the failing playwright test is just a fluke, see #689. I think we should disable that test until we figure that out.

I split off the first two commits into a separate PR: #690, so they can be independently reviewed.

not-my-profile commented:
This is ready for review ... I don't know what's causing the playwright CI fluke, see #689 ... but I also cannot easily debug that since the CIs are currently almost always failing at the prior dependencies job (#685), which is fixed by this PR.

stephenmathieson commented on .circleci/config.yml:
```diff
@@ -187,6 +128,7 @@ jobs:
       - restore_dependency_cache
       - browser-tools/install-chrome
       - browser-tools/install-chromedriver
+      - run: npm install -g npm@7
```
Why pin this at v7? I imagine this will be easy to overlook when we upgrade Node.js.

not-my-profile replied:
A few lines above that we we're using a node14 image due to WDIO. Node 14 comes with npm 6.14, which doesn't yet support NPM workspaces.

stephenmathieson commented on packages/cli/package.json:
```diff
@@ -32,8 +32,7 @@
     "build": "tsc",
     "test": "mocha --timeout 60000 -r ts-node/register 'src/**/**.test.ts'",
-    "coverage": "nyc npm run test",
-    "prepare": "npm run build"
```
Why is the `prepare` script being removed? IIRC we rely on it to transpile before publishing to npm.

straker replied:
Talked with @michael-siek and this isn't needed. When left in it tries to build the cli before it links packages together, which would cause the build to fail as @axe-core/webdriverjs wasn't available yet.

stephenmathieson commented on packages/cli/src/lib/utils.test.ts:
```diff
@@ -73,25 +72,6 @@ describe('utils', () => {
-      afterEach(() => {
-        mock.restore();
-      });
-      it('fall back to use `locally` installed axe-core', () => {
```
Can we implement a different test rather than removing this?

not-my-profile replied:
The fact that it defaults to some locally installed axe-core is already tested indirectly through several other test cases. The node_module resolution code has been simplified to `axePath = require.resolve('axe-core/axe.js');`, which I don't really think requires dedicated testing.

stephenmathieson commented on packages/cli/src/lib/utils.ts:
```diff
@@ -82,23 +82,7 @@ export const getAxeSource = (axePath?: string): string | void => {
 
   if (!fs.existsSync(axePath)) {
-    // Look for axe in CWD ./node_modules
-    axePath = path.join(process.cwd(), 'node_modules', 'axe-core', 'axe.js');
```
Should this be considered a breaking change? @michael-siek

stephenmathieson reviewed (commented):
Few small things left inline. This is looking good thought 👍 

Will let someone with more recent familiarity with this project give the final review.

michael-siek commented:
Hey @not-my-profile just reaching out to see if you've seen the request changes and merge conflicts on this PR.

not-my-profile commented:
I rebased the PR but I currently don't have time to look into the new CI failures. I split off the first four preparatory commits into #757 ... I think it makes sense to land these changes first.

Zidious commented:
> I rebased the PR but I currently don't have time to look into the new CI failures. I split off the first four preparatory commits into #757 ... I think it makes sense to land these changes first.

Hey @not-my-profile, 

When you get a moment, there are merge conflicts. Feel free to close and re-open if you don't have time at the moment. 

Thanks again!

not-my-profile commented:
If I close it, I'll probably forget that it exists. I just converted it to a draft to signal that it's not ready for review yet (due to the CI failures I mentioned earlier).

straker commented on packages/cli/package.json:
```diff
@@ -45,7 +44,7 @@
   "dependencies": {
     "@axe-core/webdriverjs": "^4.7.3",
-    "axe-core": "^4.7.0",
+    "axe-core": "~4.7.0",
```
Axe-core 4.8.1 was released and I don't want to take it with this conversion, so pinned all axe-core versions to 4.7

straker commented on packages/cli/src/bin/cli.test.ts:
```diff
@@ -391,7 +391,7 @@ describe('cli', () => {
   });
 
-  describe('--chromedriver-path', () => {
+  describe.skip('--chromedriver-path', () => {
```
This test fails due to the way workspaces manages node_modules, will open a tech debt ticket to fix it.

straker commented on packages/puppeteer/package.json:
```diff
@@ -44,7 +44,7 @@
     "mocha": "^10.0.0",
     "nyc": "^15.1.0",
-    "puppeteer": "^19.7.3",
+    "puppeteer": "19.7.3",
```
Starting in Puppeteer 19.11 (ish?) they started adding a log that stated that `headlness: new` would be needed (which massively clogs up the test run as it logs a huge paragraph every test). However, if you add `headless: new` to the options it breaks the unloaded iframe test. It seems that whatever is in `headless: new` fixes the problem with unloaded iframes? Will open a tech debt ticket to investigate.

straker replied:
https://github.com/dequelabs/axe-core-npm/issues/828

straker commented on packages/webdriverio/package.json:
```diff
@@ -70,7 +70,7 @@
     "tsup": "^6.7.0",
     "typescript": "^4.8.4",
-    "webdriverio": "^8.8.2"
+    "webdriverio": "8.8.2"
```
Latest webdriverio breaks our typescript build due to this bug with types https://github.com/webdriverio/webdriverio/issues/11055. Will open a tech debt ticket to fix it.

straker replied:
https://github.com/dequelabs/axe-core-npm/issues/829

dbjorge commented on packages/puppeteer/tsconfig.json:
```diff
@@ -1,24 +1,4 @@
-    "moduleResolution": "node",
-    "esModuleInterop": true
-  },
+  "extends": "../../tsconfig.json",
```
This is hiding a change from `module: commonjs` to `module: esnext` that probably shouldn't be part of this PR

dbjorge commented on packages/react/examples/next.js/package-lock.json:
Should these examples be added as workspaces so CI will verify they build?

straker replied:
We never had tests to run the examples so they were never verified to begin with. If we want to have them run tests we should do so in a different pr.

straker commented on packages/puppeteer/package.json:
```diff
@@ -53,10 +53,7 @@
   "dependencies": {
-    "axe-core": "^4.7.0"
-  },
-  "peerDependencies": {
```
It was not. Not sure when that happened. Good catch.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dequelabs/axe-core-npm?pullRequest=687) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/dequelabs/axe-core-npm?pullRequest=687) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: straker<br/>:x: dbjorge<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dequelabs/axe-core-npm?pullRequest=687) it.</sub>

dbjorge commented:
force-push was to correct author email address in merge from develop, to hopefully fix CLA bot

dbjorge commented:
Reviewed for security

dbjorge reviewed (dismissed):
Ah, wait, approved too early - @straker , there's a missing lockfile update to pick up the changes from 2dedf1b

michael-siek commented:
- [x] reviewed for security
